rules/firestore.json

```diff
@@ -35,50 +35,205 @@ service cloud.firestore {
+                             (safeLastUpdate(resource.data.lastUpdate, request.resource.data.lastUpdate));
+            // All verified user can read parts (needed for the lobby)
+            allow read: if userIsVerified();
+        }
+        function userIsChosenOpponentInJoiner(partId) {
+            let joiner = get(/databases/$(database)/documents/joiners/$(partId)).data;
+            return hasField(joiner, 'chosenOpponent') && joiner.chosenOpponent.id == request.auth.uid;
+        }
+        function userIsPlayer(part) {
+            let userIsPlayerZero = part.playerZero.id == request.auth.uid;
+            let userIsPlayerOne = hasField(part, 'playerOne') && part.playerOne.id == request.auth.uid;
+            return userIsPlayerZero || userIsPlayerOne;
+        }
+        function safeTurnUpdate(oldTurn, newTurn) {
+            return (newTurn == oldTurn + 1) || (newTurn == oldTurn - 1) || (newTurn == oldTurn - 2);
+        }
+        function safeLastUpdate(oldLastUpdate, newLastUpdate) {
+            let legitimatePlayerZeroUpdate = newLastUpdate.player == 0 && resource.data.playerZero.id == request.auth.uid;
+            let legitimatePlayerOneUpdate = hasField(resource.data, 'playerOne') && newLastUpdate.player == 1 && resource.data.playerOne.id == request.auth.uid;
+            let indexIncremented = newLastUpdate.index == oldLastUpdate.index + 1;
+            return (legitimatePlayerZeroUpdate || legitimatePlayerOneUpdate || partHasStarted(resource.data) == false) && indexIncremented;
+        }
+        function partHasStarted(part) {
+            // A part is started when the beginning field is set
+            return hasField(part, 'beginning');
+        }
+        function hasTimedOut(user) {
+            let userDoc =  get(/databases/$(database)/documents/users/$(user.id)).data;
+            let timedOut = duration.value(request.time.toMillis(), 'ms') > duration.value(userDoc.last_changed.seconds, 's') + duration.value(10, 's');
+            return timedOut;
+        }
+        match /joiners/{partId} {
+            // Any verified user can create a joiner as long as they are themselves creator and if the corresponding part exists
+            allow create: if userIsVerified() &&
+                             userIsCreatorWithValidName(request.resource.data) &&
+                             exists(/databases/$(database)/documents/parts/$(partId));
+            // The creator can delete a joiner if the part has not started
+            allow delete: if userIsVerified() && userIsCreator(resource.data) && joinerHasNotStarted(resource.data);
+            // Anybody else can delete a joiner only if the creator has timed out and the part has not started
+            allow delete: if userIsVerified() && hasTimedOut(creatorOf(partId)) && joinerHasNotStarted(resource.data);
+            // The creator can change all fields except changing partStatus to STARTED (3), until the part starts
```

Comment:
to STARTED(3) or FINISHED(4)
//prefer-explicit-rule :p

Reply:
tout à fait pour les deux, changé